File: swift/tapster.swift
import Foundation

class Card {
    var number: String
    var cardholderName: String
    var balance: Float = 0.0
    var ouncesPoured: Float = 0.0

    init(_ number: String, _ cardholderName: String) {
        self.number = number
        self.cardholderName = cardholderName
    }
}

enum ReaderEventType {
    case INSERTED, CHARGED, REMOVED
}

struct ReaderEvent: CustomStringConvertible {
    var eventType: ReaderEventType
    var timestamp = Int(Date().timeIntervalSince1970 * 1000)
    var payload: String

    public var description: String {
        return "\(timestamp) - \(eventType) - \(payload)"
    }
}

struct Reader {
    var currentCard: Card? = nil
    var events: [ReaderEvent] = []

    mutating func insertCard(_ card: inout Card) {
        events.append(ReaderEvent(eventType: .INSERTED, payload: card.number))
        currentCard = card
    }

    mutating func removeCard() {
        events.append(ReaderEvent(eventType: .REMOVED, payload: card.number))
        currentCard = nil
    }

    mutating func chargeCard(_ ouncesPoured: Float, _ pricePerOunce: Float) {
        guard currentCard != nil else { return }
        
        let charge = ouncesPoured * pricePerOunce
        events.append(ReaderEvent(eventType: .CHARGED, payload: String(format: "%.2f", charge)))
        currentCard!.balance += charge
        currentCard!.ouncesPoured += ouncesPoured
    }

    func displayStats() {
        if currentCard != nil {
            print("Cardholder: \(currentCard!.cardholderName)")
            print(String(format: "Total Amount: $%.2f", currentCard!.balance))
            print(String(format: "Ounces Poured: %.2f", currentCard!.ouncesPoured))
        }

        print()

        for event in events {
            print(event)
        }
    }
}

var card = Card("5555555555555555","Ray Perry")
var reader = Reader()

reader.insertCard(&card)
reader.chargeCard(10.1, 0.79)
reader.removeCard()
reader.insertCard(&card)
reader.displayStats()

File: c/tapster.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

typedef struct {
  char * number;
  char * cardholder_name;
  double balance;
  double ounces_poured;
} Card;

typedef enum {
  INSERTED = 0,
  CHARGED,
  REMOVED
} ReaderEventType;

const char * EVENT_TYPES[] = { "INSERTED", "CHARGED", "REMOVED" };

typedef struct {
  ReaderEventType type;
  int timestamp;
  char * payload;
} ReaderEvent;

typedef struct {
  void * current_card;
  ReaderEvent events[100];
  int number_of_events;
} Reader;

void insert_card(Reader * r, Card * card) {
  ReaderEvent new_event = { INSERTED, (int) time(NULL), card->cardholder_name };
  r->events[r->number_of_events] = new_event;
  r->number_of_events += 1;
  r->current_card = card;
} 

void remove_card(Reader * r) {
  ReaderEvent new_event = { REMOVED, (int) time(NULL), ((Card *) r->current_card)->cardholder_name };
  r->events[r->number_of_events] = new_event;
  r->number_of_events += 1; 
  r->current_card = NULL;
}

void charge_card(Reader * r, double ounces_poured, double price_per_ounce) {
  if (r->current_card == NULL) return;
  double charge = ounces_poured * price_per_ounce;
  char payload[32];
  sprintf(payload, "%f", charge);
  ReaderEvent new_event = { CHARGED, (int) time(NULL), payload };
  r->events[r->number_of_events] = new_event;
  r->number_of_events += 1;
  ((Card *) r->current_card)->balance += charge;
  ((Card *) r->current_card)->ounces_poured += ounces_poured;
}

void display_stats(Reader * r) {
  if (r->current_card == NULL) return;
  printf("Cardholder name: %s\n", ((Card *) r->current_card)->cardholder_name);
  printf("Balance: %.2f\n", ((Card *) r->current_card)->balance);
  printf("Ounces poured: %.2f\n", ((Card *) r->current_card)->ounces_poured);
  for (int i = 0; i < r->number_of_events; i++) {
    if (r->events[i].type == CHARGED) {
      double charge;
      sscanf(r->events[i].payload, "%lf", &charge);
      printf("%d - %s - %.2lf\n", r->events[i].timestamp, EVENT_TYPES[r->events[i].type], charge);
    } else {
      printf("%d - %s - %s\n", r->events[i].timestamp, EVENT_TYPES[r->events[i].type], r->events[i].payload);
    }
  }
}

int main(void) {
  Card card = { ((char *) "5555555555555555"), ((char *) "Ray Perry"), 0.0, 0.0 };
  Reader reader;

  insert_card(&reader, &card);
  charge_card(&reader, 10.3, 0.79);
  remove_card(&reader);
  insert_card(&reader, &card);

  display_stats(&reader);
  
  return 0;
}

// Local Variables:
// compile-command: "gcc -std=c11 -Wall -Wextra -Wpedantic -Wformat=2 -Wno-unused-parameter -Wshadow -Wwrite-strings -Wstrict-prototypes -Wold-style-definition -Wredundant-decls -Wnested-externs -Wmissing-include-dirs -o ./tapster tapster.c && ./tapster"
// End:
